bip-unknown-var-budget-script.mediawiki

```diff
@@ -0,0 +1,722 @@
+
+Arithmetic opcodes are extended to allow variable-length operands beyond 4 bytes, though they are now unsigned.  Other opcodes which interpret numbers or place them on the stack now use unsigned values.  This includes the final stack element after execution (i.e. no "negative zero").  Bit operations are re-enabled, as are multiplication and division.
+
+The individual stack entry limit of 520 bytes is replaced by a larger individual limit and a total limit.
+
+===A Model For Opcodes Dealing With Stack Data===
+
+When large stack objects are permitted, the bottleneck for script operations is based on the time taken to process the stack data it accesses (the exceptions being signature operations).
+
+All current opcodes are most reasonably implemented using linear accesses to stack data (though sometimes multiple times), so the cost model uses the length of the stack inputs (or occasionally, outputs).
+
+- We assume that the manipulation of the stack vector itself (e.g. OP_DROP) is negligible.
+- We assume an implementation uses simple linear arrays/vectors of contiguous memory.
+- Specific operations are defined as "fast", as their implementations have been empirically demonstated to be optimized: comparing bytes, comparing bytes against zero, copying bytes and zeroing bytes.  These cost 1 unit per byte.
+- SHA256 is a special operation which costs 10 units per byte.
+- Other operations are defined as "slow": these cost 2 units per byte.
+
+The budget for signature operations is subsumed into the varops budget, by simply consuming 5200 * 50 varops per CHECKSIG operation. [FIXME: After the first "free" one?]
+
+The aim is to limit worst-case script evaluation timing, without meaningfully impacting any useful script operations.  Note that an implementation which is more efficient than the versions modeled does not introduce any problems (though a future soft-fork version may want to reflect this in reduced costings): only an implementation with a signficantly worse worst-case behaviour would be problematic.
+
+In particular, costs are based on worst-case behavior (which often does not occur) and budget limits are based on maximum stack element sizes (where smaller, cache-hot elements will be significantly faster).
+
+The worst case behaviour for a previously-possible stack entry is SHA256 of 520 bytes: this costs 5200 varops budget, which could be paid for by the OP_SHA256 opcode itself.  This shows trivially that all previously-existing operations come within budget under this BIP.
+
+===Limits===
+
+The individual stack entry limit of 520 bytes is increased to the total block weight (4000000) bytes.  An additional gross limit of twice that (8000000) applies across all stack entries.  The limit of 1000 total stack entries remains.
```

Comment:
In conversation I heard that for complex scripts the stack size limit is one of the most difficult limitations to work around (see also https://bitvmx.org/knowledge/optimizing-algorithms-for-bitcoin-script "limit"). But maybe it's less of a concern in the GSR because with MUL and bit-operations there's less reason to load large lookup tables on the stack.

Reply:
Also, maybe less of a problem in the GSR because one can potentially use SUBSTR to obtain stack elements from one large stack element.